File: 202102220-java-lift-kata/src/main/java/codingdojo/Lift.java
package codingdojo;

import java.util.ArrayList;
import java.util.List;

public class Lift {
    private final String id;
    private int floor;
    private List<Integer> requests;
    private boolean doorsOpen;

    public Lift(String id, int floor, List<Integer> requests, boolean doorsOpen) {
        this.id = id;
        this.floor = floor;
        this.requests = new ArrayList<>(requests);
        this.doorsOpen = doorsOpen;
    }

    public Lift(String id, int floor) {
        this(id, floor, new ArrayList<Integer>(), false);
    }

    public Lift(String id, int floor, List<Integer> requests) {
        this(id, floor, requests, false);
    }

    public Lift(String id, int floor, boolean doorsOpen) {
        this(id, floor, new ArrayList<Integer>(), doorsOpen);
    }

    public String getId() {
        return id;
    }

    public int getFloor() {
        return floor;
    }

    public boolean hasRequestForFloor(int floor) {
        return this.requests.contains(floor);
    }

    public boolean areDoorsOpen() {
        return doorsOpen;
    }

    public void tick() {
        if (doorsOpen) {
            closeDoors();
            return;
        }

        this.floor = requests.remove(0);
        openDoors();
    }

    private void openDoors() {
        doorsOpen = true;
    }

    private void closeDoors() {
        doorsOpen = false;
    }
}

File: 20210225-kotlin-lift-kata/src/main/kotlin/codingdojo/Lift.kt
package codingdojo

class Lift(
        val id: String,
        private var floor: Int,
        private val requests: List<Int> = emptyList(),
        private var doorsOpen: Boolean = false) {

    fun getFloor() = floor

    fun hasRequestForFloor(floor: Int): Boolean {
        return requests.contains(floor)
    }

    fun areDoorsOpen(): Boolean {
        return doorsOpen
    }

    fun tick() {
        if (doorsOpen) {
            closeDoors()
            return
        }
        if (floor != requests.first()) {
            floor = requests.first()
        } else {
            doorsOpen = true
        }
    }

    private fun closeDoors() {
        doorsOpen = false
    }
}
